{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "35b3c5df",
   "metadata": {
    "tags": []
   },
   "source": [
    "### Study: \"Building an optimal predictive model for imputing tissue-specific gene expression by combining genotype and whole blood transcriptome data\"\n",
    "### Method: MERGED_fixed_100\n",
    "### Author/Maintainer: Sunwoo Jung < s17171717s@gmail.com >"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "284f8648",
   "metadata": {
    "tags": []
   },
   "source": [
    "### This code is for: \n",
    "### instructions on how to access and explore the imputaiton models generated by a method named MERGED_fixed_100."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "7f6bf599",
   "metadata": {},
   "outputs": [],
   "source": [
    "# You can SKIP this code if you located the Rdata file (sample_models_m.Rdata) and this jupyter notebook in the same directory.\n",
    "# Make sure to set your working directory to the folder that contains the Rdata file.\n",
    "# Below is the example used by the writer.\n",
    "setwd('/Users/sunwoosimac/downloads/models')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "c96c21b9",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load the data.\n",
    "load('sample_models_m.Rdata')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "a55eac8a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "'models'"
      ],
      "text/latex": [
       "'models'"
      ],
      "text/markdown": [
       "'models'"
      ],
      "text/plain": [
       "[1] \"models\""
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Data structure for the imputation models is List.\n",
    "# List named 'models' will be loaded.\n",
    "ls()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "a72218ce",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"list\"\n"
     ]
    }
   ],
   "source": [
    "print(class(models))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e343fad6",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "ef7baf08",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"The imputatioin weights for < 2000 genes > are available.\"\n"
     ]
    }
   ],
   "source": [
    "# The length of the List 'models': The number of genes included in a target tissue (Here: sample data containing only 2000 genes randomly selected from 'GTEx v7 Adipose Subcutaneous' tissue).\n",
    "N_gene = length(models) \n",
    "print(paste0('The imputatioin weights for < ', N_gene, ' genes > are available.')) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "047dc580",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"ENSG00000147642\"\n"
     ]
    }
   ],
   "source": [
    "# Each element of the List 'models' contains the imputation model for each gene of a target tissue.\n",
    "# Each element of the List 'models' is named after the corresponding gene ID.\n",
    "# For example, the first element of the List contains the imputation model(weights) for the gene ENSG00000147642.\n",
    "print(names(models)[1])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "afff34bf",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " [1] \"ENSG00000147642\" \"ENSG00000124225\" \"ENSG00000157470\" \"ENSG00000149294\"\n",
      " [5] \"ENSG00000228290\" \"ENSG00000240747\" \"ENSG00000196712\" \"ENSG00000184182\"\n",
      " [9] \"ENSG00000237298\" \"ENSG00000171824\" \"ENSG00000104133\" \"ENSG00000116649\"\n",
      "[13] \"ENSG00000102931\" \"ENSG00000196367\" \"ENSG00000186088\" \"ENSG00000059377\"\n",
      "[17] \"ENSG00000011600\" \"ENSG00000204632\" \"ENSG00000112796\" \"ENSG00000131408\"\n"
     ]
    }
   ],
   "source": [
    "# Here are some other examples.\n",
    "print(names(models)[c(1:10,1991:2000)])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "b8a42e88",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"MERGED_fixed\"\n"
     ]
    }
   ],
   "source": [
    "# Each item (imputation model) in the List is named after the corresponding method, in this case, 'MERGED_fixed'.\n",
    "print(names(models[[\"ENSG00000147642\"]]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "0a399843",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"dgCMatrix\"\n",
      "attr(,\"package\")\n",
      "[1] \"Matrix\"\n"
     ]
    }
   ],
   "source": [
    "# The stored item is the matrix that contains the weights needed for\n",
    "#     imputing the expression level of the corresponding gene. \n",
    "# This matrix of the imputation weights is our imputation model. \n",
    "print(class(models[[\"ENSG00000147642\"]]$'MERGED_fixed'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0514459a",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "3531244b",
   "metadata": {
    "tags": []
   },
   "source": [
    "### Each item in the List 'models' is the matrix of the weights needed for imputing the expression level of the corresponding gene.\n",
    "### The imputation model for the gene ENSG00000147642 can be displayed as below."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "9933c954",
   "metadata": {},
   "outputs": [],
   "source": [
    "imputation_model <- models[[\"ENSG00000147642\"]]$'MERGED_fixed'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "35f09207",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "10 x 1 sparse Matrix of class \"dgCMatrix\"\n",
       "                                 s1\n",
       "(Intercept)            6.555347e-16\n",
       "SEX                   -1.766743e-02\n",
       "AGE                   -2.763743e-02\n",
       "PC1                   -9.612259e-02\n",
       "PC2                   -5.006442e-02\n",
       "PC3                   -6.251501e-02\n",
       "PC4                    2.266458e-02\n",
       "PC5                   -1.103225e-01\n",
       "8_110086771_G_A_b37_A  6.677403e-05\n",
       "8_110087871_T_A_b37_A  6.811716e-05"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "10 x 1 sparse Matrix of class \"dgCMatrix\"\n",
       "                           s1\n",
       "ENSG00000128159 -0.0004739657\n",
       "ENSG00000100429 -0.0028413186\n",
       "ENSG00000188130 -0.0008489317\n",
       "ENSG00000185386 -0.0023584904\n",
       "ENSG00000196576 -0.0025749849\n",
       "ENSG00000100258  0.0006312531\n",
       "ENSG00000025770  0.0005358800\n",
       "ENSG00000177989 -0.0038872774\n",
       "ENSG00000130487 -0.0018919609\n",
       "ENSG00000079974  0.0001536264"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(imputation_model, 10)\n",
    "tail(imputation_model, 10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "28bc82a1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"The number of predictors: 5087\"\n"
     ]
    }
   ],
   "source": [
    "N_predictors = length(imputation_model)\n",
    "print(paste0(\"The number of predictors: \", N_predictors)) \n",
    "# For the gene ENSG00000147642, the matrix stores the imputation weights of 5087 predictors.\n",
    "# Predictors are either the covariates (sex, age, PCs), the genotypes (SNPs), or the whole blood expression."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "58564fce",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " [1] \"(Intercept)\"           \"SEX\"                   \"AGE\"                  \n",
      " [4] \"PC1\"                   \"PC2\"                   \"PC3\"                  \n",
      " [7] \"PC4\"                   \"PC5\"                   \"8_110086771_G_A_b37_A\"\n",
      "[10] \"8_110087871_T_A_b37_A\" \"8_110088888_A_G_b37_G\" \"8_110089913_T_A_b37_A\"\n",
      "[13] \"8_110090736_G_T_b37_T\" \"8_110093634_T_C_b37_C\" \"8_110093871_A_G_b37_G\"\n",
      "[16] \"8_110099209_C_T_b37_T\" \"8_110102250_A_T_b37_T\" \"8_110103774_T_C_b37_C\"\n",
      "[19] \"8_110105081_A_G_b37_G\" \"8_110105147_G_A_b37_A\" \"ENSG00000238160\"      \n",
      "[22] \"ENSG00000131437\"       \"ENSG00000164402\"       \"ENSG00000213585\"      \n",
      "[25] \"ENSG00000081059\"       \"ENSG00000113575\"       \"ENSG00000119048\"      \n",
      "[28] \"ENSG00000152700\"       \"ENSG00000113648\"       \"ENSG00000146021\"      \n",
      "[31] \"ENSG00000177733\"      \n"
     ]
    }
   ],
   "source": [
    "# The rownames of the matrix are the names of the corresponding predictors.\n",
    "print(rownames(models[[\"ENSG00000147642\"]]$'MERGED_fixed')[c(1:20,2000:2010)])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2fedadd1",
   "metadata": {},
   "source": [
    "### Note: Anyone can freely use the imputation weights stored in the file for any research purpose."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c23fcb86",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "r"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.3.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
