{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "35b3c5df",
   "metadata": {},
   "source": [
    "### Study: \"Building an optimal predictive model for imputing tissue-specific gene expression by combining genotype and whole blood transcriptome data\"\n",
    "### Method: MERGED_flexible\n",
    "### Writer/Maintainer: Sunwoo Jung < s17171717s@gmail.com >"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "284f8648",
   "metadata": {},
   "source": [
    "### This code is for: \n",
    "### Access to the imputaiton models generated by MERGED_flexible model."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "7f6bf599",
   "metadata": {},
   "outputs": [],
   "source": [
    "# You can SKIP this code if you located the Rdata file (imputation models) and this jupyter notebook in the same directory.\n",
    "# Make sure to set your working directory to the folder that contains the Rdata file.\n",
    "# Below is the example used by the writer.\n",
    "setwd('/Users/sunwoosimac/downloads/models')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "c96c21b9",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load the data.\n",
    "load('Adipose_Subcutaneous_m.Rdata')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "a55eac8a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "'models'"
      ],
      "text/latex": [
       "'models'"
      ],
      "text/markdown": [
       "'models'"
      ],
      "text/plain": [
       "[1] \"models\""
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Data structure for the imputation models is List.\n",
    "# List named 'models' will be loaded.\n",
    "ls()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "a72218ce",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"list\"\n"
     ]
    }
   ],
   "source": [
    "print(class(models))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e343fad6",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "ef7baf08",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"The imputatioin weights for < 15216 genes > are available.\"\n"
     ]
    }
   ],
   "source": [
    "# The length of the List 'models' = The number of genes included in a target tissue (Here: Adipose Subcutaneous).\n",
    "N_gene = length(models) \n",
    "print(paste0('The imputatioin weights for < ', N_gene, ' genes > are available.')) \n",
    "# The imputation models for 15216 genes are available for Adipose Subcutaneous."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "047dc580",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"ENSG00000237683\"\n"
     ]
    }
   ],
   "source": [
    "# Each element of the List contains the imputation model for each gene of a target tissue (Here: Adipose Subcutaneous).\n",
    "# Each element of the List is named as the corresponding gene ID.\n",
    "# For example, the first element of the List contains the imputation model(weights) for the gene ENSG00000237683.\n",
    "print(names(models)[1])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "afff34bf",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " [1] \"ENSG00000237683\" \"ENSG00000241860\" \"ENSG00000228463\" \"ENSG00000237094\"\n",
      " [5] \"ENSG00000235373\" \"ENSG00000228327\" \"ENSG00000237491\" \"ENSG00000177757\"\n",
      " [9] \"ENSG00000225880\" \"ENSG00000228794\" \"ENSG00000272821\" \"ENSG00000025708\"\n",
      "[13] \"ENSG00000177989\" \"ENSG00000205560\" \"ENSG00000100288\" \"ENSG00000008735\"\n",
      "[17] \"ENSG00000100299\" \"ENSG00000251322\" \"ENSG00000100312\" \"ENSG00000079974\"\n"
     ]
    }
   ],
   "source": [
    "# Here are some other examples.\n",
    "print(names(models)[c(1:10,15207:15216)])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "b8a42e88",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"1\"\n"
     ]
    }
   ],
   "source": [
    "# In each element of the List, the item (imputation model) is named as the value \n",
    "#     of the regularization ratio parameter (phi) selected for the correspondign gene in MERGED_flexible model.\n",
    "# For detailed explanation, please refer to our study.\n",
    "\n",
    "# For example, \n",
    "#     MERGED_flexible model selected '1' for the value of phi \n",
    "#     for imputing the expression level of the gene \"ENSG00000237683\". \n",
    "\n",
    "print(names(models[[\"ENSG00000237683\"]]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "a2ecafd2",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"10\"\n",
      "[1] \"100\"\n",
      "[1] \"100\"\n",
      "[1] \"0.1\"\n"
     ]
    }
   ],
   "source": [
    "# Here are some other examples.\n",
    "print(names(models[[\"ENSG00000114993\"]]))\n",
    "print(names(models[[\"ENSG00000134250\"]]))\n",
    "print(names(models[[\"ENSG00000162426\"]]))\n",
    "print(names(models[[\"ENSG00000237438\"]]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "0a399843",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"dgCMatrix\"\n",
      "attr(,\"package\")\n",
      "[1] \"Matrix\"\n"
     ]
    }
   ],
   "source": [
    "# In each element of the List, the stored item is the matrix that contains the weights needed for\n",
    "#     imputing the expression level of the corresponding gene. \n",
    "# This matrix of the imputation weights is our imputation model. \n",
    "print(class(models[[\"ENSG00000100299\"]]$'1'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0514459a",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "3531244b",
   "metadata": {},
   "source": [
    "### Each item of the List 'models' is the matrix of the weights needed for imputing the expression level of the corresponding gene.\n",
    "### The imputation model for the gene ENSG00000100299 is displayed below."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "9933c954",
   "metadata": {},
   "outputs": [],
   "source": [
    "imputation_model <- models[[\"ENSG00000100299\"]]$'1'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "35f09207",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "필요한 패키지를 로딩중입니다: Matrix\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "6 x 1 sparse Matrix of class \"dgCMatrix\"\n",
       "                       s1\n",
       "(Intercept)  2.766291e-18\n",
       "SEX          6.683749e-02\n",
       "AGE          9.659364e-02\n",
       "PC1         -8.409612e-02\n",
       "PC2         -8.724955e-03\n",
       "PC3          1.802840e-02"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "6 x 1 sparse Matrix of class \"dgCMatrix\"\n",
       "                           s1\n",
       "ENSG00000100304 -0.0011983941\n",
       "ENSG00000100364 -0.0004302000\n",
       "ENSG00000197182 -0.0034459036\n",
       "ENSG00000205643 -0.0013794376\n",
       "ENSG00000260708 -0.0041834980\n",
       "ENSG00000188511  0.0009948689"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(imputation_model)\n",
    "tail(imputation_model)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "28bc82a1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"The number of predictors: 3518\"\n"
     ]
    }
   ],
   "source": [
    "N_predictors = length(imputation_model)\n",
    "print(paste0(\"The number of predictors: \", N_predictors)) \n",
    "# For the gene ENSG00000100299, the impuation weights of 3518 predictors are stored in the matrix.\n",
    "# Predictors are either the genotype (SNPs) or the whole blood expression, along with the covariates (sex, age, PCs)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "58564fce",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " [1] \"(Intercept)\"             \"SEX\"                    \n",
      " [3] \"AGE\"                     \"PC1\"                    \n",
      " [5] \"PC2\"                     \"PC3\"                    \n",
      " [7] \"PC4\"                     \"PC5\"                    \n",
      " [9] \"22_50561364_G_A_b37_A\"   \"22_50561417_A_G_b37_G\"  \n",
      "[11] \"22_50561590_G_A_b37_A\"   \"22_50562311_G_C_b37_C\"  \n",
      "[13] \"22_50564145_T_A_b37_A\"   \"22_50564238_G_A_b37_A\"  \n",
      "[15] \"22_50564380_C_T_b37_T\"   \"22_50564419_C_T_b37_T\"  \n",
      "[17] \"22_50564442_C_T_b37_T\"   \"22_50564510_T_C_b37_C\"  \n",
      "[19] \"22_50564618_A_AT_b37_AT\" \"22_50565583_A_G_b37_G\"  \n",
      "[21] \"ENSG00000130347\"         \"ENSG00000130349\"        \n",
      "[23] \"ENSG00000081087\"         \"ENSG00000112335\"        \n",
      "[25] \"ENSG00000135535\"         \"ENSG00000112367\"        \n",
      "[27] \"ENSG00000112290\"         \"ENSG00000168438\"        \n",
      "[29] \"ENSG00000004809\"         \"ENSG00000155115\"        \n",
      "[31] \"ENSG00000196591\"        \n"
     ]
    }
   ],
   "source": [
    "# The rownames of the matrix are the names of the corresponding predictors.\n",
    "print(rownames(models[[\"ENSG00000100299\"]]$'1')[c(1:20,2000:2010)])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2fedadd1",
   "metadata": {},
   "source": [
    "### Anyone can freely use the imputation weights stored in the file for any research purpose."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c23fcb86",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "r"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.2.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
